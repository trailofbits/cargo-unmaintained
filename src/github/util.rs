use anyhow::{Context, Result, anyhow};
use std::{
    env::var,
    fs::{File, OpenOptions, create_dir_all, read_to_string},
    io::{Write, stdin},
    path::PathBuf,
    sync::{LazyLock, OnceLock},
};

#[allow(clippy::unwrap_used)]
static CONFIG_DIRECTORY: LazyLock<PathBuf> = LazyLock::new(|| {
    #[cfg(not(windows))]
    {
        let base_directories = xdg::BaseDirectories::new().unwrap();
        base_directories.get_config_file("cargo-unmaintained")
    }
    #[cfg(windows)]
    {
        let local_app_data = var("LOCALAPPDATA").unwrap();
        PathBuf::from(local_app_data).join("cargo-unmaintained")
    }
});

static TOKEN_PATH: LazyLock<PathBuf> = LazyLock::new(|| CONFIG_DIRECTORY.join("token.txt"));

pub(super) static PERSONAL_TOKEN: OnceLock<String> = OnceLock::new();

pub fn load_token(f: impl FnOnce(&str) -> Result<()>) -> Result<bool> {
    let token_untrimmed = if let Ok(path) = var("GITHUB_TOKEN_PATH") {
        read_to_string(&path).with_context(|| format!("failed to read {path:?}"))?
    } else if let Ok(token) = var("GITHUB_TOKEN") {
        // smoelius: Suppress warning if `CI` is set, i.e., if running on GitHub.
        if var("CI").is_err() {
            #[cfg(__warnings)]
            crate::warn!(
                "found a token in `GITHUB_TOKEN`; consider using the more secure method of \
                 setting `GITHUB_TOKEN_PATH` to the path of a file containing the token",
            );
        }
        token
    } else if TOKEN_PATH.try_exists().with_context(|| {
        format!(
            "failed to determine whether `{}` exists",
            TOKEN_PATH.display()
        )
    })? {
        read_to_string(&*TOKEN_PATH).with_context(|| format!("failed to read {TOKEN_PATH:?}"))?
    } else {
        #[cfg(__warnings)]
        crate::warn!(
            "`GITHUB_TOKEN_PATH` and `GITHUB_TOKEN` are not set and no file was found at {}; \
             archival statuses will not be checked",
            TOKEN_PATH.display()
        );
        return Ok(false);
    };
    let token = token_untrimmed.trim_end().to_owned();
    PERSONAL_TOKEN
        .set(token.clone())
        .map_err(|_| anyhow!("`load_token` was already called"))?;
    f(&token)?;
    Ok(true)
}

pub(crate) fn save_token() -> Result<()> {
    println!("Please paste a personal access token below. The token needs no scopes.");

    let mut buf = String::new();

    {
        let n = stdin()
            .read_line(&mut buf)
            .with_context(|| "failed to read stdin")?;
        assert_eq!(buf.len(), n);
    }

    create_dir_all(&*CONFIG_DIRECTORY).with_context(|| "failed to create config directory")?;

    let mut file = OpenOptions::new()
        .create(true)
        .truncate(true)
        .write(true)
        .open(&*TOKEN_PATH)
        .with_context(|| format!("failed to open `{}`", TOKEN_PATH.display()))?;
    set_permissions(&file, 0o600)?;
    file.write_all(buf.as_bytes())
        .with_context(|| format!("failed to write `{}`", TOKEN_PATH.display()))?;

    println!(
        "Personal access token written to `{}`",
        TOKEN_PATH.display()
    );

    Ok(())
}

type CargoResult<T> = Result<T>;

// smoelius: The below definitions of `set_permissions` were copied from:
// https://github.com/rust-lang/cargo/blob/1e6828485eea0f550ed7be46ef96107b46aeb162/src/cargo/util/config.rs#L1010-L1024
#[cfg(unix)]
#[cfg_attr(dylint_lib = "try_io_result", allow(try_io_result))]
fn set_permissions(file: &File, mode: u32) -> CargoResult<()> {
    use std::os::unix::fs::PermissionsExt;

    let mut perms = file.metadata()?.permissions();
    perms.set_mode(mode);
    file.set_permissions(perms)?;
    Ok(())
}

#[cfg(not(unix))]
#[allow(unused, clippy::unnecessary_wraps)]
fn set_permissions(file: &File, mode: u32) -> CargoResult<()> {
    Ok(())
}
