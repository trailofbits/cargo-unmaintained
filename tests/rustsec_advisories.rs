use snapbox::assert_matches_path;
use std::{env::remove_var, process::Command};

mod util;
use util::{tee, token_modifier, Tee};

#[ctor::ctor]
fn initialize() {
    remove_var("CARGO_TERM_COLOR");
}

#[test]
fn rustsec_advisories() {
    let mut command = Command::new("cargo");
    command
        .args(["run", "--bin=rustsec_advisories"])
        .current_dir("rustsec_util")
        .env("RUST_BACKTRACE", "0");

    let output = tee(command, Tee::Stdout).unwrap();

    let stdout_actual = std::str::from_utf8(&output.captured).unwrap();

    assert_matches_path(
        format!("tests/rustsec_advisories.{}.stdout", token_modifier()),
        stdout_actual,
    );
}
