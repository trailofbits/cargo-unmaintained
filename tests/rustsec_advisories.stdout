136 advisories for unmaintained packages
lz4-compress...found
serial...not found
tempdir...found
typemap...not found
boxfnonce...found
net2...found
block-cipher-trait...error:
```
    Updating crates.io index
error: failed to select a version for the requirement `block-cipher-trait = "*"`
  version 0.1.0 is yanked
  version 0.2.0 is yanked
  version 0.3.0 is yanked
  version 0.4.0 is yanked
  version 0.4.1 is yanked
  version 0.4.2 is yanked
  version 0.5.0 is yanked
  version 0.5.1 is yanked
  version 0.5.2 is yanked
  version 0.5.3 is yanked
  version 0.6.0 is yanked
  version 0.6.1 is yanked
  version 0.6.2 is yanked
  version 0.7.0 is yanked
  version 0.99.0 is yanked
location searched: crates.io index
required by package `block-cipher-trait-temp-package v0.1.0 ([..])`
```
stb_truetype...not found
bigint...found
failure...found
block-cipher...error:
```
Error: found no packages matching `block-cipher`
```
stream-cipher...error:
```
Error: found no packages matching `stream-cipher`
```
safe-nd...found
ffi_utils...found
fake_clock...found
safe_bindgen...found
quic-p2p...found
routing...found
safe_app...error:
```
    Updating crates.io index
error: failed to select a version for the requirement `clap = "~2.14.0"`
  version 2.14.0 is yanked
  version 2.14.1 is yanked
location searched: crates.io index
required by package `resource_proof v0.4.0`
    ... which satisfies dependency `resource_proof = "~0.4.0"` of package `routing v0.31.0`
    ... which satisfies dependency `routing = "~0.31.0"` of package `safe_app v0.1.0`
    ... which satisfies dependency `safe_app = "*"` of package `safe_app-temp-package v0.1.0 ([..])`
```
safe_authenticator...error:
```
    Updating crates.io index
error: failed to select a version for the requirement `clap = "~2.14.0"`
  version 2.14.0 is yanked
  version 2.14.1 is yanked
location searched: crates.io index
required by package `resource_proof v0.4.0`
    ... which satisfies dependency `resource_proof = "~0.4.0"` of package `routing v0.31.0`
    ... which satisfies dependency `routing = "~0.31.0"` of package `safe_authenticator v0.1.0`
    ... which satisfies dependency `safe_authenticator = "*"` of package `safe_authenticator-temp-package v0.1.0 ([..])`
```
safe_vault...error:
```
Error: found no packages matching `safe_vault`
```
safe_core...found
stderr...found
rulinalg...found
slice-deque...not found
tokio-proto...found
term_size...found
mach...not found
interfaces2...found
safe-api...error:
```
    Updating crates.io index
error: failed to select a version for the requirement `quinn = "~0.3.0"`
  version 0.3.0 is yanked
location searched: crates.io index
required by package `quic-p2p v0.2.1`
    ... which satisfies dependency `quic-p2p = "~0.2.1"` of package `safe_core v0.33.0`
    ... which satisfies dependency `safe_core = "~0.33.0"` of package `safe-api v0.4.0`
    ... which satisfies dependency `safe-api = "*"` of package `safe-api-temp-package v0.1.0 ([..])`
```
jsonrpc-quic...found
office...found
aesni...found
aes-soft...found
aes-ctr...found
miscreant...error:
```
    Updating crates.io index
error: failed to select a version for the requirement `miscreant = "*"`
  version 0.0.0 is yanked
  version 0.1.0 is yanked
  version 0.2.0 is yanked
  version 0.3.0 is yanked
  version 0.4.0 is yanked
  version 0.4.1 is yanked
  version 0.4.2 is yanked
  version 0.5.0 is yanked
  version 0.5.1 is yanked
  version 0.5.2 is yanked
  version 0.99.0 is yanked
location searched: crates.io index
required by package `miscreant-temp-package v0.1.0 ([..])`
```
cpuid-bool...found
anymap...leaf
spirv_headers...found
cosmos_sdk...found
serde_cbor...found
cargo-download...error:
```
Error: found no packages matching `cargo-download`
```
rental...found
sass-rs...found
sodiumoxide...found
ansi_term...not found
rusttype...found
dotenv...not found
dotenv_codegen...not found
traitobject...leaf
twoway...not found
nphysics3d...found
nphysics2d...found
ncollide3d...found
ncollide2d...found
encoding...not found
lmdb...found
ftd2xx-embedded-hal...found
pty...found
static_type_map...found
double-checked-cell...not found
r2d2_odbc...found
odbc...found
markdown...not found
interledger-packet...found
mapr...found
wee_alloc...found
clipboard...found
badge...found
orbtk...found
parity-wasm...found
rusoto_credential...found
alloc-cortex-m...found
claim...found
parity-util-mem...found
json...leaf
kuchiki...found
git-hash...found
git-path...found
buf_redux...found
tree_magic...found
xsalsa20poly1305...found
users...not found
ftp...found
tui...found
multipart...found
dlopen_derive...found
fehler...not found
cpython...not found
safemem...found
hpack...found
loopdev...found
cosmwasm...error:
```
    Updating crates.io index
error: failed to select a version for the requirement `cosmwasm = "*"`
  version 0.1.0 is yanked
  version 0.2.0 is yanked
  version 0.3.0 is yanked
  version 0.4.0 is yanked
  version 0.4.1 is yanked
  version 0.4.2 is yanked
  version 0.5.0 is yanked
  version 0.5.1 is yanked
  version 0.5.2 is yanked
  version 0.6.0 is yanked
  version 0.6.1 is yanked
  version 0.6.2 is yanked
  version 0.6.3 is yanked
  version 0.6.4 is yanked
  version 0.7.0 is yanked
  version 0.7.1 is yanked
  version 0.7.2 is yanked
location searched: crates.io index
required by package `cosmwasm-temp-package v0.1.0 ([..])`
```
generational-arena...found
filesystem...found
yaml-rust...not found
puccinier...error:
```
Error: found no packages matching `puccinier`
```
rsa-export...found
libp2p-tokio-socks5...found
zip_next...found
proc-macro-error...found
atty...recently updated
pqcrypto-dilithium...found
pqcrypto-kyber...found
hwloc...found
bcc...found
instant...not found
cw0...found
strason...found
opentelemetry_api...found
derivative...found
openslide...found
minitrace...recently updated
mmap...found
chrono-english...not found
conrod_core...found
conrod...found
gdkwayland...found
gdkwayland-sys...found
gdk...found
atk...found
gdkx11-sys...found
gtk...found
atk-sys...found
gdkx11...found
gdk-sys...found
gtk3-macros...found
gtk-sys...found
gtk-layer-shell...found
gtk-layer-shell-sys...found
get-size...not found
get-size-derive...found
paste...found
openpgp-card-sequoia...found
backoff...found
resolve...found
not found - error (11)
    block-cipher-trait    https://rustsec.org/advisories/RUSTSEC-2020-0018.html
    block-cipher          https://rustsec.org/advisories/RUSTSEC-2020-0057.html
    stream-cipher         https://rustsec.org/advisories/RUSTSEC-2020-0058.html
    safe_app              https://rustsec.org/advisories/RUSTSEC-2020-0083.html
    safe_authenticator    https://rustsec.org/advisories/RUSTSEC-2020-0084.html
    safe_vault            https://rustsec.org/advisories/RUSTSEC-2020-0085.html
    safe-api              https://rustsec.org/advisories/RUSTSEC-2021-0024.html
    miscreant             https://rustsec.org/advisories/RUSTSEC-2021-0062.html
    cargo-download        https://rustsec.org/advisories/RUSTSEC-2021-0133.html
    cosmwasm              https://rustsec.org/advisories/RUSTSEC-2024-0004.html
    puccinier             https://rustsec.org/advisories/RUSTSEC-2024-0331.html
not found - leaf (3)
    anymap                https://rustsec.org/advisories/RUSTSEC-2021-0065.html
    traitobject           https://rustsec.org/advisories/RUSTSEC-2021-0144.html
    json                  https://rustsec.org/advisories/RUSTSEC-2022-0081.html
not found - recently-updated (2)
    atty                  https://rustsec.org/advisories/RUSTSEC-2024-0375.html
    minitrace             https://rustsec.org/advisories/RUSTSEC-2024-0390.html
not found - other (19)
    serial                https://rustsec.org/advisories/RUSTSEC-2017-0008.html
    typemap               https://rustsec.org/advisories/RUSTSEC-2019-0039.html
    stb_truetype          https://rustsec.org/advisories/RUSTSEC-2020-0020.html
    slice-deque           https://rustsec.org/advisories/RUSTSEC-2020-0158.html
    mach                  https://rustsec.org/advisories/RUSTSEC-2020-0168.html
    ansi_term             https://rustsec.org/advisories/RUSTSEC-2021-0139.html
    dotenv                https://rustsec.org/advisories/RUSTSEC-2021-0141.html
    dotenv_codegen        https://rustsec.org/advisories/RUSTSEC-2021-0142.html
    twoway                https://rustsec.org/advisories/RUSTSEC-2021-0146.html
    encoding              https://rustsec.org/advisories/RUSTSEC-2021-0153.html
    double-checked-cell   https://rustsec.org/advisories/RUSTSEC-2022-0024.html
    markdown              https://rustsec.org/advisories/RUSTSEC-2022-0044.html
    users                 https://rustsec.org/advisories/RUSTSEC-2023-0040.html
    fehler                https://rustsec.org/advisories/RUSTSEC-2023-0067.html
    cpython               https://rustsec.org/advisories/RUSTSEC-2023-0076.html
    yaml-rust             https://rustsec.org/advisories/RUSTSEC-2024-0320.html
    instant               https://rustsec.org/advisories/RUSTSEC-2024-0384.html
    chrono-english        https://rustsec.org/advisories/RUSTSEC-2024-0395.html
    get-size              https://rustsec.org/advisories/RUSTSEC-2024-0425.html
found (101)
    lz4-compress          https://rustsec.org/advisories/RUSTSEC-2017-0007.html
    tempdir               https://rustsec.org/advisories/RUSTSEC-2018-0017.html
    boxfnonce             https://rustsec.org/advisories/RUSTSEC-2019-0040.html
    net2                  https://rustsec.org/advisories/RUSTSEC-2020-0016.html
    bigint                https://rustsec.org/advisories/RUSTSEC-2020-0025.html
    failure               https://rustsec.org/advisories/RUSTSEC-2020-0036.html
    safe-nd               https://rustsec.org/advisories/RUSTSEC-2020-0063.html
    ffi_utils             https://rustsec.org/advisories/RUSTSEC-2020-0064.html
    fake_clock            https://rustsec.org/advisories/RUSTSEC-2020-0065.html
    safe_bindgen          https://rustsec.org/advisories/RUSTSEC-2020-0066.html
    quic-p2p              https://rustsec.org/advisories/RUSTSEC-2020-0067.html
    routing               https://rustsec.org/advisories/RUSTSEC-2020-0076.html
    safe_core             https://rustsec.org/advisories/RUSTSEC-2020-0086.html
    stderr                https://rustsec.org/advisories/RUSTSEC-2020-0109.html
    rulinalg              https://rustsec.org/advisories/RUSTSEC-2020-0147.html
    tokio-proto           https://rustsec.org/advisories/RUSTSEC-2020-0162.html
    term_size             https://rustsec.org/advisories/RUSTSEC-2020-0163.html
    interfaces2           https://rustsec.org/advisories/RUSTSEC-2021-0002.html
    jsonrpc-quic          https://rustsec.org/advisories/RUSTSEC-2021-0025.html
    office                https://rustsec.org/advisories/RUSTSEC-2021-0034.html
    aesni                 https://rustsec.org/advisories/RUSTSEC-2021-0059.html
    aes-soft              https://rustsec.org/advisories/RUSTSEC-2021-0060.html
    aes-ctr               https://rustsec.org/advisories/RUSTSEC-2021-0061.html
    cpuid-bool            https://rustsec.org/advisories/RUSTSEC-2021-0064.html
    spirv_headers         https://rustsec.org/advisories/RUSTSEC-2021-0096.html
    cosmos_sdk            https://rustsec.org/advisories/RUSTSEC-2021-0099.html
    serde_cbor            https://rustsec.org/advisories/RUSTSEC-2021-0127.html
    rental                https://rustsec.org/advisories/RUSTSEC-2021-0134.html
    sass-rs               https://rustsec.org/advisories/RUSTSEC-2021-0136.html
    sodiumoxide           https://rustsec.org/advisories/RUSTSEC-2021-0137.html
    rusttype              https://rustsec.org/advisories/RUSTSEC-2021-0140.html
    nphysics3d            https://rustsec.org/advisories/RUSTSEC-2021-0148.html
    nphysics2d            https://rustsec.org/advisories/RUSTSEC-2021-0149.html
    ncollide3d            https://rustsec.org/advisories/RUSTSEC-2021-0150.html
    ncollide2d            https://rustsec.org/advisories/RUSTSEC-2021-0151.html
    lmdb                  https://rustsec.org/advisories/RUSTSEC-2022-0001.html
    ftd2xx-embedded-hal   https://rustsec.org/advisories/RUSTSEC-2022-0005.html
    pty                   https://rustsec.org/advisories/RUSTSEC-2022-0015.html
    static_type_map       https://rustsec.org/advisories/RUSTSEC-2022-0023.html
    r2d2_odbc             https://rustsec.org/advisories/RUSTSEC-2022-0036.html
    odbc                  https://rustsec.org/advisories/RUSTSEC-2022-0039.html
    interledger-packet    https://rustsec.org/advisories/RUSTSEC-2022-0050.html
    mapr                  https://rustsec.org/advisories/RUSTSEC-2022-0053.html
    wee_alloc             https://rustsec.org/advisories/RUSTSEC-2022-0054.html
    clipboard             https://rustsec.org/advisories/RUSTSEC-2022-0056.html
    badge                 https://rustsec.org/advisories/RUSTSEC-2022-0057.html
    orbtk                 https://rustsec.org/advisories/RUSTSEC-2022-0060.html
    parity-wasm           https://rustsec.org/advisories/RUSTSEC-2022-0061.html
    rusoto_credential     https://rustsec.org/advisories/RUSTSEC-2022-0071.html
    alloc-cortex-m        https://rustsec.org/advisories/RUSTSEC-2022-0073.html
    claim                 https://rustsec.org/advisories/RUSTSEC-2022-0077.html
    parity-util-mem       https://rustsec.org/advisories/RUSTSEC-2022-0080.html
    kuchiki               https://rustsec.org/advisories/RUSTSEC-2023-0019.html
    git-hash              https://rustsec.org/advisories/RUSTSEC-2023-0025.html
    git-path              https://rustsec.org/advisories/RUSTSEC-2023-0026.html
    buf_redux             https://rustsec.org/advisories/RUSTSEC-2023-0028.html
    tree_magic            https://rustsec.org/advisories/RUSTSEC-2023-0036.html
    xsalsa20poly1305      https://rustsec.org/advisories/RUSTSEC-2023-0037.html
    ftp                   https://rustsec.org/advisories/RUSTSEC-2023-0043.html
    tui                   https://rustsec.org/advisories/RUSTSEC-2023-0049.html
    multipart             https://rustsec.org/advisories/RUSTSEC-2023-0050.html
    dlopen_derive         https://rustsec.org/advisories/RUSTSEC-2023-0051.html
    safemem               https://rustsec.org/advisories/RUSTSEC-2023-0081.html
    hpack                 https://rustsec.org/advisories/RUSTSEC-2023-0084.html
    loopdev               https://rustsec.org/advisories/RUSTSEC-2023-0088.html
    generational-arena    https://rustsec.org/advisories/RUSTSEC-2024-0014.html
    filesystem            https://rustsec.org/advisories/RUSTSEC-2024-0015.html
    rsa-export            https://rustsec.org/advisories/RUSTSEC-2024-0333.html
    libp2p-tokio-socks5   https://rustsec.org/advisories/RUSTSEC-2024-0334.html
    zip_next              https://rustsec.org/advisories/RUSTSEC-2024-0337.html
    proc-macro-error      https://rustsec.org/advisories/RUSTSEC-2024-0370.html
    pqcrypto-dilithium    https://rustsec.org/advisories/RUSTSEC-2024-0380.html
    pqcrypto-kyber        https://rustsec.org/advisories/RUSTSEC-2024-0381.html
    hwloc                 https://rustsec.org/advisories/RUSTSEC-2024-0382.html
    bcc                   https://rustsec.org/advisories/RUSTSEC-2024-0383.html
    cw0                   https://rustsec.org/advisories/RUSTSEC-2024-0385.html
    strason               https://rustsec.org/advisories/RUSTSEC-2024-0386.html
    opentelemetry_api     https://rustsec.org/advisories/RUSTSEC-2024-0387.html
    derivative            https://rustsec.org/advisories/RUSTSEC-2024-0388.html
    openslide             https://rustsec.org/advisories/RUSTSEC-2024-0389.html
    mmap                  https://rustsec.org/advisories/RUSTSEC-2024-0394.html
    conrod_core           https://rustsec.org/advisories/RUSTSEC-2024-0396.html
    conrod                https://rustsec.org/advisories/RUSTSEC-2024-0397.html
    gdkwayland            https://rustsec.org/advisories/RUSTSEC-2024-0410.html
    gdkwayland-sys        https://rustsec.org/advisories/RUSTSEC-2024-0411.html
    gdk                   https://rustsec.org/advisories/RUSTSEC-2024-0412.html
    atk                   https://rustsec.org/advisories/RUSTSEC-2024-0413.html
    gdkx11-sys            https://rustsec.org/advisories/RUSTSEC-2024-0414.html
    gtk                   https://rustsec.org/advisories/RUSTSEC-2024-0415.html
    atk-sys               https://rustsec.org/advisories/RUSTSEC-2024-0416.html
    gdkx11                https://rustsec.org/advisories/RUSTSEC-2024-0417.html
    gdk-sys               https://rustsec.org/advisories/RUSTSEC-2024-0418.html
    gtk3-macros           https://rustsec.org/advisories/RUSTSEC-2024-0419.html
    gtk-sys               https://rustsec.org/advisories/RUSTSEC-2024-0420.html
    gtk-layer-shell       https://rustsec.org/advisories/RUSTSEC-2024-0422.html
    gtk-layer-shell-sys   https://rustsec.org/advisories/RUSTSEC-2024-0423.html
    get-size-derive       https://rustsec.org/advisories/RUSTSEC-2024-0427.html
    paste                 https://rustsec.org/advisories/RUSTSEC-2024-0436.html
    openpgp-card-sequoia  https://rustsec.org/advisories/RUSTSEC-2025-0011.html
    backoff               https://rustsec.org/advisories/RUSTSEC-2025-0012.html
    resolve               https://rustsec.org/advisories/RUSTSEC-2025-0013.html
---
As of 2025-03-07, the RustSec Advisory Database contains 136 active advisories for unmaintained packages. Using the above conditions, `cargo-unmaintained` automatically identifies 101 (74%) of them. These results can be reproduced by running the [`rustsec_advisories`] example within this repository.
- Of the 35 packages in the RustSec Advisory Database _not_ identified by `cargo-unmaintained`:
  - 11 do not build
  - 3 are existent, unarchived leaves
  - 2 were updated within the past 365 days
  - 19 were not identified for other reasons
